### DISCUSSION TITLE
Python version ?

### DISCUSSION THREAD

**Discussion by moodyblue** [BlazejosP/huawei-sun2000-API-CLI#25] [Ideas]
Are there any plans to have a python version of this tool ?

**Comment by BlazejosP** [comment]
I never think about this before but of course I can consider this but can you explain to me what will be an advantage in comprehension to this what we have now  and why Python version may be some help to you?

**Comment by moodyblue** [comment]
I would like to use this tool in a computer running 24x7. Currently I have a NAS and a router running linux embedded distros where httpie is not available. But python3 is installed and I can use it to access internet providers. Here is one that I'm using `https://github.com/dgomes/pyprosegur`.

IMO a python version can be used in a wider range of equipments.

**Reply by BlazejosP** [reply]
What do you think If I ask others on forum who will be interested in python version and if we can have some collaborators?

**Reply by moodyblue** [reply]
That would be wonderfull.

**Comment by BlazejosP** [comment]
Thanks for sharing this info we me. I can say that I understand way that why be usefully especially on specialised Linux boxes where is impossible to install tools by apt-get or other package manager. I also understand that python3 is more perspective in some future than good old bash. Probably on the beginning you and other people my need just CLI interface TUI is not necessary on small router box where  you configure everything and make this just working 24x7.

So to have this what you need done will be necessary to translate especially functions.sh  and CLI interface fusionsolarapp.sh rest is not important to made this working. But that will be not easy translation which we can make just with automatics tool. Because  I also put inside this scripts some external linux tools calls so will be necessary to found some python functions/library's which can do something similarly or add code which can handle that.
My plans for now are finishing all the functions in TUI polishing a little bit application and add back possibility to send  to MQTT and influxDB-Grafana after that will be possible to use send data to Domoticz/Home Asistant

After that I can start rewriting code to Python so realistically I can work on this in spring probably code may be ready for summer because I working on that in my spare time. But if you or others persons will be interested to speed up this process I think we need more collaborators and we can divide task amongst  different persons propably each function to one person and than that my go faster. Meaby lets see first if will be others also interested in having a python version of Huawei sun2000 CLI?

**Comment by januweness** [comment]
Hi, I could not support as I am completely Python illiterate. Since Python has it's own issues with libraries and versions, I don't think it will be trivial...



----- Original Message -----
From: BlazejosP ***@***.***>

Subject: Re: [BlazejosP/huawei-sun2000-API-CLI] Python version ? (Discussion #25)

What do you think If I ask others on forum who will be interested in python version and if we can have some collaborators?


This message is intended only for the recipient(s) named above. It may contain proprietary information and/or
protected content. Any unauthorised disclosure, use, retention or dissemination is prohibited. If you have received
this e-mail in error, please notify the sender immediately. ESA applies appropriate organisational measures to protect
personal data, in case of data privacy queries, please contact the ESA Data Protection Officer ***@***.***).

**Reply by BlazejosP** [reply]
I myself know python meaby not very well but as for now have some experience. Improving ginlog scraper which propably I will upload.
@januweness if such pythonised version will be created in next year will have some use for you for example?

**Comment by januweness** [comment]
Hi, you ask whether a Python version would be useful (for me): yes and no: 
For me personally I could run a Python version of course but not modify it. However, more relevant is the general usefulness which I see divided: 
Yes: because there is a larger Python audience out there than a bash audience 
No: because I estimate the Python audience is more demanding in terms of documentation than bash developers 
Looking at what we have achieved with our resources (all spare time!) I am not sure we can achieve the degree of documentation that would be expected. One has to budget for documentation about the same amount of time as needed for development! 

I personally appreciated the bash scripts because I can customize them to my own needs. There is probably no such person as a "bash user" but only bash developers who tend to take what they get, write a few wrapper functions and get done with it. Once we release a Python version, we will be confronted with "users" who might expect a lot more description, fancy interfaces etc. We may then attract a lot of support questions ..... 

Perhaps we could create a Python wrapper that calls up the bash code? That would of course limit it to Unix and MAC platforms but maybe that would be enough? Not sure how many Python users work under Windows.....

**Reply by BlazejosP** [reply]
Thanks for that point @januweness  that really opens eyes on problems which lay ahead if I go this path. What do you think   @moodyblue  Python wrapper with bash code inside will be some solution in your case if you have limited Linux box even if that is some type of  quick&dirty solution?

**Reply by moodyblue** [reply]
> Python wrapper with bash code inside

I don't know. I'm not a technical guy. Would this solve the problem of not having `httpie` installed ?

Probably a docker version of this tool would be easier to implement and it will also reach a wider audience.

**Reply by BlazejosP** [reply]
good point with this docker version as I understand you have an possibility to run docker packages?

**Comment by moodyblue** [comment]
BlazejosP ***@***.***> escreveu em qua, 17/11/2021 às 17:13 :

> good point with this docker version as I understand you have an
> possibility to run docker packages?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BlazejosP/huawei-sun2000-API-CLI/discussions/25#discussioncomment-1659533>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABOK2D7MXJI3727SUIMFTVTUMPPBXANCNFSM5HWMZ4SQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
Yes I can, although I don't like it

**Reply by BlazejosP** [reply]
What about an snap container I was able to build one yesterday and made working with Sun2000-API-CLI is any possibility that you will be able to run snap on your system?

**Comment by BlazejosP** [comment]
Hello I now consider two options to help you in this situation:

1.  using https://www.swag.uwaterloo.ca/bash2py/index.html bash2py to translate automatically our bash program to python and then try fine tune hard parts like using of external programs or some parts which automatics will fail.
2. building docker image 

To be honest with both haven't much experience that will take some time even until late spring
As for now I try research what can be done faster

**Comment by BlazejosP** [comment]
What about snap containers I made this app working inside as a experiment already?
